// nebb-rest-service/bundles/nebb-rest-service.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs'), require('rxjs/operators'), require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('nebb-rest-service', ['exports', 'rxjs', 'rxjs/operators', '@angular/core'], factory) :
    (factory((global['nebb-rest-service'] = {}),global.rxjs,global.rxjs.operators,global.ng.core));
}(this, (function (exports,rxjs,operators,core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    // unsupported: template constraints.
    /**
     * @template T
     */
    var  
    // unsupported: template constraints.
    /**
     * @template T
     */
    ResourceService = (function () {
        function ResourceService(httpClient, url, endpoint, serializer) {
            this.httpClient = httpClient;
            this.url = url;
            this.endpoint = endpoint;
            this.serializer = serializer;
        }
        /**
         * @param {?} item
         * @return {?}
         */
        ResourceService.prototype.create = /**
         * @param {?} item
         * @return {?}
         */
            function (item) {
                return this.httpClient
                    .post(this.url + "/" + this.endpoint, this.serializer.toJson(item))
                    .pipe(operators.catchError(this.handleError('create', item)));
            };
        /**
         * @param {?} item
         * @return {?}
         */
        ResourceService.prototype.update = /**
         * @param {?} item
         * @return {?}
         */
            function (item) {
                return this.httpClient
                    .put(this.url + "/" + this.endpoint + "/" + item.id, this.serializer.toJson(item))
                    .pipe(operators.catchError(this.handleError('update', item)));
            };
        /**
         * @param {?} id
         * @return {?}
         */
        ResourceService.prototype.get = /**
         * @param {?} id
         * @return {?}
         */
            function (id) {
                return this.httpClient.get(this.url + "/" + this.endpoint + "/" + id);
            };
        /**
         * @param {?} queryOptions
         * @return {?}
         */
        ResourceService.prototype.list = /**
         * @param {?} queryOptions
         * @return {?}
         */
            function (queryOptions) {
                var _this = this;
                return this.httpClient
                    .get(this.url + "/" + this.endpoint + "?" + queryOptions.toQueryString())
                    .pipe(operators.tap(function (t) { return _this.log("fetched data"); }));
            };
        /**
         * @param {?} id
         * @return {?}
         */
        ResourceService.prototype.delete = /**
         * @param {?} id
         * @return {?}
         */
            function (id) {
                var _this = this;
                return this.httpClient.delete(this.url + "/" + this.endpoint + "/" + id).pipe(operators.tap(function (t) { return _this.log("fetched data"); }), operators.catchError(this.handleError('delete')));
            };
        /**
         * @param {?=} operation
         * @param {?=} result
         * @return {?}
         */
        ResourceService.prototype.handleError = /**
         * @param {?=} operation
         * @param {?=} result
         * @return {?}
         */
            function (operation, result) {
                var _this = this;
                if (operation === void 0) {
                    operation = 'operation';
                }
                return function (error) {
                    console.error(error);
                    _this.log(operation + " failed: " + error.message);
                    return rxjs.of(/** @type {?} */ (result));
                };
            };
        /**
         * @param {?} message
         * @return {?}
         */
        ResourceService.prototype.log = /**
         * @param {?} message
         * @return {?}
         */
            function (message) {
                console.log('Resource.service: ' + message);
            };
        return ResourceService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Resource = (function () {
        function Resource(values) {
            if (values === void 0) {
                values = {};
            }
            Object.assign(this, values);
        }
        return Resource;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NebbRestServiceModule = (function () {
        function NebbRestServiceModule() {
        }
        NebbRestServiceModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [],
                        exports: []
                    },] },
        ];
        return NebbRestServiceModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var QueryOptions = (function () {
        function QueryOptions() {
            this.pageNumber = 1;
            this.pageSize = 10000;
        }
        /**
         * @return {?}
         */
        QueryOptions.prototype.toQueryMap = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ queryMap = new Map();
                queryMap.set('pageNumber', "" + this.pageNumber);
                queryMap.set('pageSize', "" + this.pageSize);
                return queryMap;
            };
        /**
         * @return {?}
         */
        QueryOptions.prototype.toQueryString = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ queryString = '';
                this.toQueryMap().forEach(function (value, key) {
                    queryString = queryString.concat(key + "=" + value + "&");
                });
                return queryString.substring(0, queryString.length - 1);
            };
        return QueryOptions;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.ResourceService = ResourceService;
    exports.Resource = Resource;
    exports.NebbRestServiceModule = NebbRestServiceModule;
    exports.QueryOptions = QueryOptions;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
