// nebb-rest-service/esm5/lib/query-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var QueryOptions = /** @class */ (function () {
    function QueryOptions() {
        this.pageNumber = 1;
        this.pageSize = 10000;
    }
    /**
     * @return {?}
     */
    QueryOptions.prototype.toQueryMap = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ queryMap = new Map();
        queryMap.set('pageNumber', "" + this.pageNumber);
        queryMap.set('pageSize', "" + this.pageSize);
        return queryMap;
    };
    /**
     * @return {?}
     */
    QueryOptions.prototype.toQueryString = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ queryString = '';
        this.toQueryMap().forEach(function (value, key) {
            queryString = queryString.concat(key + "=" + value + "&");
        });
        return queryString.substring(0, queryString.length - 1);
    };
    return QueryOptions;
}());
export { QueryOptions };
function QueryOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    QueryOptions.prototype.pageNumber;
    /** @type {?} */
    QueryOptions.prototype.pageSize;
}
