// nebb-rest-service/fesm2015/nebb-rest-service.js
import { of } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// unsupported: template constraints.
/**
 * @template T
 */
class ResourceService {
    /**
     * @param {?} httpClient
     * @param {?} url
     * @param {?} endpoint
     * @param {?} serializer
     */
    constructor(httpClient, url, endpoint, serializer) {
        this.httpClient = httpClient;
        this.url = url;
        this.endpoint = endpoint;
        this.serializer = serializer;
    }
    /**
     * @param {?} item
     * @return {?}
     */
    create(item) {
        return this.httpClient
            .post(`${this.url}/${this.endpoint}`, this.serializer.toJson(item))
            .pipe(catchError(this.handleError('create', item)));
    }
    /**
     * @param {?} item
     * @return {?}
     */
    update(item) {
        return this.httpClient
            .put(`${this.url}/${this.endpoint}/${item.id}`, this.serializer.toJson(item))
            .pipe(catchError(this.handleError('update', item)));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    get(id) {
        return this.httpClient.get(`${this.url}/${this.endpoint}/${id}`);
    }
    /**
     * @param {?} queryOptions
     * @return {?}
     */
    list(queryOptions) {
        return this.httpClient
            .get(`${this.url}/${this.endpoint}?${queryOptions.toQueryString()}`)
            .pipe(tap(t => this.log(`fetched data`)));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    delete(id) {
        return this.httpClient.delete(`${this.url}/${this.endpoint}/${id}`).pipe(tap(t => this.log(`fetched data`)), catchError(this.handleError('delete')));
    }
    /**
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    handleError(operation = 'operation', result) {
        return (error) => {
            console.error(error);
            this.log(`${operation} failed: ${error.message}`);
            return of(/** @type {?} */ (result));
        };
    }
    /**
     * @param {?} message
     * @return {?}
     */
    log(message) {
        console.log('Resource.service: ' + message);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class Resource {
    /**
     * @param {?=} values
     */
    constructor(values = {}) {
        Object.assign(this, values);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NebbRestServiceModule {
}
NebbRestServiceModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [],
                exports: []
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class QueryOptions {
    constructor() {
        this.pageNumber = 1;
        this.pageSize = 10000;
    }
    /**
     * @return {?}
     */
    toQueryMap() {
        const /** @type {?} */ queryMap = new Map();
        queryMap.set('pageNumber', `${this.pageNumber}`);
        queryMap.set('pageSize', `${this.pageSize}`);
        return queryMap;
    }
    /**
     * @return {?}
     */
    toQueryString() {
        let /** @type {?} */ queryString = '';
        this.toQueryMap().forEach((value, key) => {
            queryString = queryString.concat(`${key}=${value}&`);
        });
        return queryString.substring(0, queryString.length - 1);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { ResourceService, Resource, NebbRestServiceModule, QueryOptions };
