// nebb-rest-service/esm2015/lib/resource.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { of } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
// unsupported: template constraints.
/**
 * @template T
 */
export class ResourceService {
    /**
     * @param {?} httpClient
     * @param {?} url
     * @param {?} endpoint
     * @param {?} serializer
     */
    constructor(httpClient, url, endpoint, serializer) {
        this.httpClient = httpClient;
        this.url = url;
        this.endpoint = endpoint;
        this.serializer = serializer;
    }
    /**
     * @param {?} item
     * @return {?}
     */
    create(item) {
        return this.httpClient
            .post(`${this.url}/${this.endpoint}`, this.serializer.toJson(item))
            .pipe(catchError(this.handleError('create', item)));
    }
    /**
     * @param {?} item
     * @return {?}
     */
    update(item) {
        return this.httpClient
            .put(`${this.url}/${this.endpoint}/${item.id}`, this.serializer.toJson(item))
            .pipe(catchError(this.handleError('update', item)));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    get(id) {
        return this.httpClient.get(`${this.url}/${this.endpoint}/${id}`);
    }
    /**
     * @param {?} queryOptions
     * @return {?}
     */
    list(queryOptions) {
        return this.httpClient
            .get(`${this.url}/${this.endpoint}?${queryOptions.toQueryString()}`)
            .pipe(tap(t => this.log(`fetched data`)));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    delete(id) {
        return this.httpClient.delete(`${this.url}/${this.endpoint}/${id}`).pipe(tap(t => this.log(`fetched data`)), catchError(this.handleError('delete')));
    }
    /**
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    handleError(operation = 'operation', result) {
        return (error) => {
            console.error(error);
            this.log(`${operation} failed: ${error.message}`);
            return of(/** @type {?} */ (result));
        };
    }
    /**
     * @param {?} message
     * @return {?}
     */
    log(message) {
        console.log('Resource.service: ' + message);
    }
}
function ResourceService_tsickle_Closure_declarations() {
    /** @type {?} */
    ResourceService.prototype.httpClient;
    /** @type {?} */
    ResourceService.prototype.url;
    /** @type {?} */
    ResourceService.prototype.endpoint;
    /** @type {?} */
    ResourceService.prototype.serializer;
}
